### ISSUE TITLE
`GoTrueClient` Memory Leak

### ISSUE THREAD

**Issue by ghost** [supabase/supabase-js#1711]
# TLDR
This line is causing a memory leak:
https://github.com/supabase/gotrue-js/blob/15c7c8258b2d42d3378be4f7738c728a07523579/src/GoTrueClient.ts#L2040

`setInterval` will persist even if the object calling it (in this case, `GoTrueClient`) has been dereferenced. This prevents garbage collection (as `setInterval` is still holding a reference, even if the user has lost all references) and leads to the memory leak.

A solution seems to have been attempted, but it is not working in (at the very least) React Native, where I encountered this issue:
https://github.com/supabase/gotrue-js/blob/15c7c8258b2d42d3378be4f7738c728a07523579/src/GoTrueClient.ts#L2043

# See Also

- the issue confirmed: https://stackoverflow.com/questions/41429030/does-setinterval-inside-an-object-prevent-garbage-collection-when-the-object-is
- potential solution: https://github.com/nbarrow-inspire-labs/bug-demo/blob/main/src/IntervalTimer.ts
  - blocked in React Native: https://github.com/facebook/react-native/issues/42742
- docs: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WeakRef

# History

I originally opened this here (https://github.com/supabase/supabase-js/issues/983) but moved once I discovered the `setInterval` issue. 

> # Bug report
> 
> - [X] I confirm this is a bug with Supabase, not with my own application. 
> - [X] I confirm I have searched the [Docs](https://docs.supabase.com), GitHub [Discussions](https://github.com/supabase/supabase/discussions), and [Discord](https://discord.supabase.com). 
> 
> ## Describe the bug
> 
> Calling `createClient` results in a `GoTrueClient` memory leak on React Native. It may be the entire `SupabaseClient`, but I only tested with `GoTrueClient` after noticing massive memory issues. See demo repo and video below. 
> 
> EDIT!: I have tracked the issue down to this line of code: 
> https://github.com/supabase/gotrue-js/blob/15c7c8258b2d42d3378be4f7738c728a07523579/src/GoTrueClient.ts#L2040
> 
> It looks like setInterval persists even after an object is dereferenced, avoiding garbage collection. Can we transfer this to @supabase/gotrue-js?
> 
> ## To Reproduce
> 
> I created a demo repo [here](https://github.com/nbarrow-inspire-labs/bug-demo). **This is an extremely simplistic example**; in our production app, we are using a router to navigate and noticed that (navigating between pages) we had memory leaks and supabase clients persisting, even though the page was no longer even being rendered and the supabase client should have been destroyed. I could not post our production app publicly, so I made this simple demo to show how even though I no longer have **any** reference to the client in my code (since it is created and destroyed after every button press), the auth client still exists and sends ticks in the background.
> 
> **NOTE**: as seen in the video below, after 2 button presses (and about 10 seconds) you can see log messages from two supabase clients (`GoTrueClient@0` and `GoTrueClient@1`). These clients are not referenced anywhere; they were (a) created on a button press, then (b) all reference to them was lost. Garbage collection would usually take care of this (I have let it run for several minutes at a time and still had references; I've ever tried creating hundreds, losing the reference, and all hundreds of them persist for seemingly without end).
> 
> Edit: I let the app run, without touching it at all, while writing this bug report (the same one shown in video, uninterrupted) and I was still seeing logs from `GoTrueClient@1` (and `@0`) at `2024-02-28T01:43:36.273Z`, which is quite a while after in terms of waiting for some kind of garbage collection to have kicked in (assuming there was **not** a memory leak, but it clearly seems like there is). Here is a log dump showing the clients persisting for quite some time (untill the process was killed): [logs.txt](https://github.com/supabase/supabase-js/files/14427571/logs.txt)
> 
> ## Expected behavior
> 
> Supabase clients should not be persisting when dereferenced; somehow garbage collection is being distorted.
> 
> ## Screenshots
> 
> https://github.com/supabase/supabase-js/assets/127700939/665eba59-b076-426a-801c-3407a2b01b03
> 
> 
> ## System information
> 
> - Version of supabase-js: `^2.39.7`

**Comment by uncvrd** [comment]
I'm pretty sure I'm dealing with something similar in Node.JS land. I just migrated my app to Supabase and my memory usage now climbs steadily over time until i get an OOM

<img width="631" alt="image" src="https://github.com/supabase/supabase-js/assets/25097099/6ac552c0-e632-4341-bc0a-547156adca8e">

Around March 9th, i updated my app to only call `supabase.auth.getUser();` when necessary instead of on each request. As you can see the slope of memory usage decreased (but since i still need to call this method for authenticated endpoints, the memory usage still grows...just slower). The steep drops are from app restarts.

I see spikes when there are large batches of `supabase.auth.getUser()` called. Running a memory heap from a load test shows that `getItemAsync()` and `_acquireLock()` are the biggest offenders. I can provide this heap if necessary. Anyone else running in to this problem serverside?

**Comment by uncvrd** [comment]
@nbarrow-inspire-labs oh thank you for leading me to the debug option! Yikes, yes, i ran a load test for 45 seconds and it has created 4000+ clients it seems like

<img width="1169" alt="image" src="https://github.com/supabase/supabase-js/assets/25097099/d52c4869-5795-4553-8c89-fc4651ba77ce">

**Comment by uncvrd** [comment]
Just wanted to report back but my hunch led me down to a memory issue with unidici, the fetch library used by node in newer versions. There were some mentions of memory issues with undici in node v18. I've updated to v20 on April 2nd and while my memory has increased a little bit initially, it seems to have somewhat stabilized, so maybe that was my issue...

<img width="765" alt="image" src="https://github.com/supabase/supabase-js/assets/25097099/01db4653-80c3-4eac-b7a9-5eeb50d4bac0">

**Comment by f1yingbanana** [comment]
Supabase auth still causes memory leak by default in the latest version!

The recommended way of initializing Supabase client on the server does not pass autoRefreshToken option, and this defaults to true, meaning that for EVERY SINGLE REQUEST we create a Supabase client that never gets garbage collected because of the [refresh loop](https://github.com/supabase/supabase-js/blob/a5f09cf9a0a8c2744464a8505333ab3136e3f290/packages/core/auth-js/src/GoTrueClient.ts#L5045):

```typescript
  private async _handleVisibilityChange() {
    if (!isBrowser() || !window?.addEventListener) {
      if (this.autoRefreshToken) {
        // in non-browser environments the refresh token ticker runs always
        this.startAutoRefresh()
      }

      return false
    }
    // ...
  }
```

This is a non-issue for serverless edge functions, but for persistent servers you will get a memory leak!

To avoid this, you must initialize the Supabase client on your server by doing:

```typescript
  const supabase = createServerClient(
    process.env.NEXT_PUBLIC_SUPABASE_URL,
    process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY,
    {
      auth: {
        autoRefreshToken: false,
      },
      cookies: {
        // ...
      },
    },
```

I don't understand why this is default true on servers. I can understand we need to refresh the token ONCE in middleware, but why would we have a loop to refresh the token?
